RFC.md

```diff
@@ -0,0 +1,466 @@
+				EarlyFusionModel
+			fusion_layer.py
+			fusion_embed.py
+			fusion_utils.py
+				register_fusion_module
+				get_fusion_params
+				set_trainable_params
+```
+
+**Pseudo code demonstrating usage of fused models**
+```python
+def flamingo_builder():
+	return DeepFusionModel(
+		encoder=CLIP(...),
+		decoder=Llama3(...))
+
+>>> model = flamingo_builder()
+
+# Load full checkpoints
+>>> model.load_state_dict(...)
+
+# Or load pretrained individual models
+>>> model.encoder.load_state_dict(...)
+>>> model.decoder.load_state_dict(...)
+
+# set only the fusion parameters to be trainable
+>>> fusion_params = get_fusion_params(model)
+>>> set_trainable_params(model, fusion_params)
+
+# Do inference with the model the same as a text model
+# Fusion models have the same signatures as TransformerDeocder
+>>> data = flamingo_transform(...) # Message formated text + images
+>>> model(**data)
+tensor([-1.0885, ..., .0231], grad_fn=<ViewBackward0>)
+```
+
+> [!Note]
+> This RFC proposes a complete design for supporting fusion models, but not all of the design is required to be built to support any one given fusion style model. This RFC gives us an opportunity to take a holistic view over the space, but it is recommended that each module/component is built as need.
+### Fusion Models
+DeepFusion requires that Decoder is already modified to have any additional layers for accepting encoder inputs (these should be wrapped with FusionLayer to be able to track these additional parameters and work with checkpoints). For inference all of the cache can be managed by the decoder, so these methods are simply passed through.
+
+```python
+class DeepFusionModel(nn.Module):
+	def __init__(self,
+		decoder: TransformerDecoder,
+		encoder: nn.Module,
+		...
+	):
+
+	def setup_caches(self, batch_size: int, dtype: torch.dtype) -> None:
+        """Setup key value caches for attention calculation.
+
+        Args:
+            batch_size (int): batch size for the caches.
+            dtype (torch.dtype): dtype for the caches.
+        """
+        self.decoder.setup_caches(batch_size, dtype)
+
+	def caches_are_enabled(self) -> bool:
+        """Check if the key value caches are setup."""
+        return self.decoder.caches_are_enabled()
+
+	def reset_caches(self):
+        """Reset the key value caches."""
+        self.decoder.reset_caches()
+
+	def forward(
+		self,
+        tokens: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+    ) -> Tensor:
+	if 	encoder_input is not None:
+		# during decoding, the output of the encoder
+		# is cached in the decoder cross attention kv_cache.
+		# so encoder_input is only called when there's a new image
+		encoder_embed = self.encoder(**encoder_input)
+	# input_pos is tracked by kv_cache now, but left in for
+	# now to support sample packing and can be used here
```

Comment:
I don't 100% follow this comment, like if we are supporting sample packing we need to have this, right?